Add new topic: 'Making an Eval command'

chroventer opened the pull request:
Added a new topic to the miscellaneous category: 'Making an Eval command'

This pull-request implements or completes the [Add a page on eval item](https://github.com/discordjs/guide/projects/1#card-16181109).

Furthermore, if you want to request any changes, feel free to do so.

GamesProSeif commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,83 @@
+    message.channel.send(clean(res), { code: 'xl' });
+} catch (err) {
+    message.channel.send(`\`Error\` \`\`\`xl\n${clean(err)}\n\`\`\``);
+}
```
Everything looks great here. Really good explanation of benefits and drawbacks of eval. Especially the securing part.
One exception though. I don't know whether it would be better not jump from Cleaning eval section to Conclusion, without providing any context of how Util#inspect works, and general explanation.

chroventer replied:
Thanks for the suggestion!
As I mentioned, I am open to any suggestions;
I'll try to improve the topic by explaining how `Util#inspect()` works in the next commit.

chroventer commented:
> I think it would be better to explain most of the stuff happening in the conclusion first.

Thanks again for the suggestion! I've updated the topic and added an explanation on how `util.inspect()` works.

Slydelix commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,95 @@
+
+### What is eval?
+The eval() function evaluates JavaScript code represented as a string.
+For instance, `eval(1 + 1)` will return `4` and `eval(client.user.tag)` will return the bot's tag (`MyBot#0001`).
```
return `2`*?

tipakA commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,110 @@
+Furthermore, anyone with the access to it can get the app's secrets and credentials - such as bot token, API keys, so on.
+
+:::warning
+Note: eval() is also slower than the alternatives, since it has to invoke the JS interpreter, while many other constructs are optimized by modern JS engines.
```
What alternatives?

tipakA commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,110 @@
+### Object to String representation
+
+We can convert object to string representation by using an in-built Node.js module - `util`. Specifically, `util#inspect()`
+So what is `util#inspect`? `util.inspect()` method returns a string representation of object that is intended for debugging.
```
Nitpick: 3 different ways to mention `Util#inspect` :P

chroventer replied:
Will replace in new commit and thanks.

tipakA commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,110 @@
+if (message.author.id !== owner) return;
+try {
+    let res = eval(args.join(' '));
+    if (typeof res !== 'string') res = inspect(res);
```
Not sure about default depth personally. It defaults to 2, and that is a lot, considering that we can reach and list all guild's information like name and icon from inspecting `message.member`. That drastically decreases chances of getting anything back except API Error about exceeding message length.

almostSouji commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,109 @@
+In other words, eval is dangerous if somehow someone else has access to it, the person can run (malicious) code with administrator access, i.e. anyone with access to it can get the app's secrets and credentials such as bot token, API keys, and so on.
+
+:::warning
+Note: eval() is also slower than the alternatives since it has to invoke the JS interpreter, while many other constructs are optimized by modern JS engines.
```
Code in code blocks, as detailed above
```suggestion
Note: `eval()` is also slower than the alternatives since it has to invoke the JS interpreter, while many other constructs are optimized by modern JS engines.
```

almostSouji commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,109 @@
+### Securing eval
+
+We will be limiting the eval command only to the owner.
+To do that, we will add a condition, which will return if the author's ID is not equals to the owner's ID.
```
As per CONTRIBUTING.md, "we" refers to the guide writers, not including the reader.
In my opinion you can directly say "To limit the eval command to the owner **you** can..."

chroventer replied:
- [x] As per CONTRIBUTING.md, "we" refers to the guide writers, not including the reader.
In my opinion, you can directly say "To limit the eval command to the owner you can..."

almostSouji commented on guide/miscellaneous/making-an-eval-command.md:
````diff
@@ -0,0 +1,109 @@
+if (message.author.id !== 'ownerID') return;
+```
+
+Additionally, you can add your ID in the config file:
````
Alternatively, not additionally, doing the same check twice is not really the best idea
```suggestion
Alternatively you can add your ID in the config file:
```

chroventer replied:
Updated!

almostSouji commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,109 @@
+
+### Cleaning eval
+
+We'll be creating a function that prevents the use of actual mentions within the return line by adding a zero-width character between the @ and the first character of the mention - blocking the mention from happening.
```
You create, we suggest, as per CONTRIBUTING.md

chroventer replied:
Done!

almostSouji commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,109 @@
+
+We can convert an object to string representation by using an in-built Node.js module - `util`. Specifically, `util#inspect()`
+So what is `util#inspect`? The `util.inspect()` method returns a string representation of an object that is intended for debugging.
+Read more on [Node.js API Docs](https://nodejs.org/api/util.html#util_util_inspect_object_options).
```
- [ ] you, we write the guide, we don't do anything else
- [ ] the entire idea behind inspect is to **not** have it be displayed as the objects string representation but rather display the object structure with its properties
- [ ] there's a bunch of doubling in 68/69 which should be merged into one coherent logic flow

chroventer replied:
- [x] the entire idea behind inspect is to not have it be displayed as the objects' string representation but rather display the object structure with its properties.
- [x] there's a bunch of doubling in 68/69 which should be merged into one coherent logic flow.
- [x] you, we write the guide, we don't do anything else

almostSouji commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,109 @@
+So what is `util#inspect`? The `util.inspect()` method returns a string representation of an object that is intended for debugging.
+Read more on [Node.js API Docs](https://nodejs.org/api/util.html#util_util_inspect_object_options).
+
+First, requiring the `util` module, however, we just need the inspect module, so we will simply destruct it from the module.
```
- [ ] I think this sentence could be greatly improved with more body to it.
"First, requiring the `util` module, however,..." doesn't read very well
- [ ] The concept is "destructuring", we don't destruct things, usually
Could also use a inline link to the guide to read more about es6 syntax: <https://discordjs.guide/additional-info/es6-syntax.html#destructuring>

chroventer replied:
- [x]  I think this sentence could be greatly improved with more body to it.
"First, requiring the util module, however,..." doesn't read very well

almostSouji commented on guide/miscellaneous/making-an-eval-command.md:
````diff
@@ -0,0 +1,109 @@
+const { inspect } = require('util');
+```
+
+Second, checking if the result is a string or not, if it is, inspect it:
````
Same as above, very short, rather confusing to read sentence structure with lots of `,`

chroventer commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,109 @@
+
+## Resulting code
+
+<resulting-code />
```
Done!

chroventer commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,109 @@
+
+### Why eval can be dangerous?
+
+eval() is a dangerous function, which executes the supplied code, with the privileges of the caller. If you run eval() with a string that could be affected by a malicious party, you may end up running malicious code on your machine. More importantly, a third-party code can see the scope in which eval() was invoked, which can lead to possible attacks.
```
- [x] Danger blocks for dangerous things, code blocks for code references (eval()).

chroventer replied:
- [x]  I don't think it's more important that someone may read your token than evaluating arbitrary code on your host machine.

chroventer replied:
- [x] "a code" is grammatically wrong

chroventer commented:
@almostSouji, please review it once again, I've made some changes.

Darqam commented on guide/miscellaneous/making-an-eval-command.md:
```diff
@@ -0,0 +1,123 @@
+`eval()` is also slower than the alternatives since it has to invoke the JS interpreter, while many other constructs are optimized by modern JS engines.
+
+::: tip
+`eval()` is also slower than the alternatives since it has to invoke the JS interpreter, while many other constructs are optimized by modern JS engines.
```
I think you have an accidental duplication there

chroventer replied:
Thanks for the suggestion, @Darqam!
